rcl/src/rcl/security_directory.c

```
@@ -246,12 +249,22 @@ const char * rcl_get_secure_root(
       "SECURITY ERROR: unable to find a folder matching the node name in %s%s."
       "Lookup strategy: %s",
       ros_secure_root_env, node_namespace, lookup_strategy);
+    allocator.deallocate(node_secure_root, allocator.state);
```

This code is introducing a lots of duplication, could we clean that up?

---

use `osrf_testing_tools_cpp` here.

---

@xabxx You can't use testing code in prod code paths. Instead, you can use the following pattern:

```c++
void my_function_that_may_throw() {}

void my_function() {
  allocate();
  try {
    my_function_that_may_throw() {};
    unallocate();
  } catch(...) {
    unallocate();
    throw;
  }
}
```

---

This is actually simpler in C because you can ignore exception-safety:

```c
Res my_function_that_may_throw() {}

void my_function() {
  allocate();
  Res res = my_function_that_may_throw() {};
  unallocate();
  return res;
}
```

---

could you help me understand what `my_function` does in this case?

---

I saw https://github.com/ros2/rcutils/blob/master/test/test_filesystem.cpp is using the testing tool and I was going to use it the same way.

---

`test_filesystem.cpp` is a test file, and C++, that's why it can use `osrf_testing_tools_cpp` and this file cannot.

---

@wjwwood @thomas-moulard I left this comment in the wrong place. Going to use `osrf_testing_tools_cpp` to deallocate memory in `test_security_directory`.

---

Going to refactor the logic based on @thomas-moulard 's suggestions.